pypsa/networkclustering.py

```diff
@@ -508,11 +503,132 @@ def busmap_by_rectangular_grid(buses, divisions=10):
+
+    if buses_i is None:
+        buses_i = network.buses.index
+
+    if feature is None:
+        logger.warning(
+            "No feature is specified for Hierarchical Clustering. "
+            "Falling back to default, where all buses have equal similarity. "
+            "You can specify a feature as pandas.DataFrame indexed with buses_i."
+        )
+
+        feature = pd.DataFrame(index=buses_i, columns=[""], data=0)
+
+    buses_x = network.buses.index.get_indexer(buses_i)
+
+    A = network.adjacency_matrix(branch_components=["Line", "Link"]).tocsc()[buses_x][:, buses_x]
+
+    labels = HAC(n_clusters=n_clusters,
+                 connectivity=A,
+                 affinity=affinity,
+                 linkage=linkage).fit_predict(feature)
+
+    busmap = pd.Series(labels, index=buses_i, dtype=str)
+
+    return busmap
+
+def hac_clustering(network, n_clusters, buses_i=None, branch_components=["Line", "Link"], feature=None,
+                   affinity='euclidean', linkage='ward', line_length_factor=1.0, **kwargs):
+    """
+    Cluster the network using Hierarchical Agglomerative Clustering.
+
+    Parameters
+    ----------
+    network : pypsa.Network
+        The buses must have coordinates x,y.
+    buses_i: None | pandas.Index, default=None
+        Subset of buses to cluster. If None, all buses are considered.
+    branch_components: List, default=["Line", "Link"]
+        Subset of all branch_components in the network.
+    feature: None | pandas.DataFrame, default=None
+        Feature to be considered for the clustering.
+        The DataFrame must be indexed with buses_i.
+        If None, all buses have the same similarity.
+    affinity: str or callable, default=’euclidean’
+        Metric used to compute the linkage.
+        Can be “euclidean”, “l1”, “l2”, “manhattan”, “cosine”, or “precomputed”.
+        If linkage is “ward”, only “euclidean” is accepted.
+        If “precomputed”, a distance matrix (instead of a similarity matrix) is needed as input for the fit method.
+    linkage: ‘ward’, ‘complete’, ‘average’ or ‘single’, default=’ward’
+        Which linkage criterion to use.
+        The linkage criterion determines which distance to use between sets of observation.
+        The algorithm will merge the pairs of cluster that minimize this criterion.
+        - ‘ward’ minimizes the variance of the clusters being merged.
+        - ‘average’ uses the average of the distances of each observation of the two sets.
+        - ‘complete’ or ‘maximum’ linkage uses the maximum distances between all observations of the two sets.
+        - ‘single’ uses the minimum of the distances between all observations of the two sets.
+    line_length_factor: float, default=1.0
+        Factor to multiply the spherical distance between two new buses in order to get new line lengths.
+    kwargs:
+        Any remaining arguments to be passed to Hierarchical Clustering (e.g. memory, connectivity).
```

Comment:
Ah, just saw that the kwargs are not passed to the HAC function. Also, perhaps it would be more clever to set the the default of `branch_components` to `n.branch_components`?

Reply:
@FabianHofmann thanks!